flight/mcu/lib/HistoryBuffer/HistoryBuffer.h

```diff
@@ -0,0 +1,144 @@
+
+#include <ChRt.h>
+
+
+/*
+History Fifo Buffer
+*/
+template <size_t size>
+class HistoryBuffer{
+    private:
+        float arr[size];
+        systime_t timestampArr[size];
+
+        int frontIndex;
+        int endIndex;
+    
+    public:
+        HistoryBuffer(){
+            frontIndex = 0;
+            endIndex = size - 1;
+        }
+
+        void push(float val, float timestamp){
+            //sets the value at the end of the queue to be equal to the desired push value
+            arr[endIndex] = val;
+            timestampArr[endIndex] = timestamp;
+
+            //since our newest value is at the end index, it becomes our start index
+            frontIndex = endIndex;
+
+            //sets the new end index to the next oldest value which will be one less than the previous end index
+            //if its currently 0, it needs to wrap around to the back
+            if(endIndex == 0){
+                endIndex = size-1;
+            }
+            else{
+                endIndex--;
+            }
+        }
+
+        float getCurrentAverage(){
+            float count = 0.0;
+            //want to average the first 50 values (start at 0, end at 49)
+            for(unsigned i = 0; i < size/2; i++){
+                //add the value of the queue at "i"th index from the front index (frontIndex + i)
+                //if our index goes over, we mod it with the length to get it to wrap back around
+                count += arr[(frontIndex + i)%size];
+            }
+            return count/(size/2);
+        }
+
+        float getPastAverage(){
+            float count = 0.0;
+            //want to average the last 50 values (start at 50, end at 99)
+            for(unsigned i = size/2; i < size; i++){
+                //add the value of the queue at "i"th index from the front index (frontIndex + i)
+                //if our index goes over, we mod it with the length to get it to wrap back around
+                count += arr[(frontIndex + i)%size];
+            }
+            return count/(size/2);
+        }
+
+        float getCurrentSecondDerivativeAverage(){
+            //we want to get velocity values from the altitude, which means that we need the change in altitidue over the change in time
+            //we can calculate the change in velocity between the current value and the next value and divide it by the time 
+            //the array will be 50-1 in length
+            float firstDiff[(size/2)-1];
+            for(unsigned i = 0; i < (size/2)-1; i++){
+                //we do arr[index] - arr[index+1] because the lower index is going to be the most recent value, which is why it goes first
+                int currentIndex = (frontIndex + i)%size;
+                int pastIndex = (frontIndex + i+1)%size;
+
+                if(timestampArr[currentIndex]-timestampArr[pastIndex] != 0){
+                    firstDiff[i] = (arr[currentIndex] - arr[pastIndex])/((timestampArr[currentIndex]-timestampArr[pastIndex])*0.001);
```

Comment:
The timestamps are pushed to the historyBuffer with the units of ticks, so we would have to use TIME_I2MS() to convert to milliseconds, and then convert to seconds.